pkg/conversion/capi2mapi/aws.go

```diff
@@ -0,0 +1,492 @@
+type machineAndAWSMachineAndAWSCluster struct {
+	machine    *capiv1.Machine
+	awsMachine *capav1.AWSMachine
+	awsCluster *capav1.AWSCluster
+}
+
+// machineSetAndAWSMachineTemplateAndAWSCluster stores the details of a Cluster API MachineSet and AWSMachineTemplate and AWSCluster.
+type machineSetAndAWSMachineTemplateAndAWSCluster struct {
+	machineSet *capiv1.MachineSet
+	template   *capav1.AWSMachineTemplate
+	awsCluster *capav1.AWSCluster
+	*machineAndAWSMachineAndAWSCluster
+}
+
+// FromMachineAndAWSMachineAndAWSCluster wraps a CAPI Machine and CAPA AWSMachine and CAPA AWSCluster into a capi2mapi MachineAndInfrastructureMachine.
+func FromMachineAndAWSMachineAndAWSCluster(m *capiv1.Machine, am *capav1.AWSMachine, ac *capav1.AWSCluster) MachineAndInfrastructureMachine {
+	return &machineAndAWSMachineAndAWSCluster{machine: m, awsMachine: am, awsCluster: ac}
+}
+
+// FromMachineSetAndAWSMachineTemplateAndAWSCluster wraps a CAPI MachineSet and CAPA AWSMachineTemplate and CAPA AWSCluster into a capi2mapi MachineSetAndAWSMachineTemplateAndAWSCluster.
+func FromMachineSetAndAWSMachineTemplateAndAWSCluster(ms *capiv1.MachineSet, mts *capav1.AWSMachineTemplate, ac *capav1.AWSCluster) MachineSetAndMachineTemplate {
+	return &machineSetAndAWSMachineTemplateAndAWSCluster{
+		machineSet: ms,
+		template:   mts,
+		awsCluster: ac,
+		machineAndAWSMachineAndAWSCluster: &machineAndAWSMachineAndAWSCluster{
+			machine: &capiv1.Machine{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      ms.Spec.Template.Labels,
+					Annotations: ms.Spec.Template.Annotations,
+				},
+				Spec: ms.Spec.Template.Spec,
+			},
+			awsMachine: &capav1.AWSMachine{
+				Spec: mts.Spec.Template.Spec,
+			},
+			awsCluster: ac,
+		},
+	}
+}
+
+// ToProviderSpec converts a capi2mapi MachineAndAWSMachineTemplateAndAWSCluster into a MAPI AWSMachineProviderConfig.
+//
+//nolint:funlen
+func (m machineAndAWSMachineAndAWSCluster) toProviderSpec() (*mapiv1.AWSMachineProviderConfig, []string, error) {
+	if m.machine == nil || m.awsMachine == nil || m.awsCluster == nil {
+		return nil, nil, errCAPIMachineAWSMachineAWSClusterCannotBeNil
+	}
+
+	var (
+		warnings []string
+		errors   field.ErrorList
+	)
+
+	fldPath := field.NewPath("spec")
+
+	mapaTenancy, err := convertAWSTenancyToMAPI(fldPath.Child("tenancy"), m.awsMachine.Spec.Tenancy)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	mapiAWSMetadataOptions, warn, errs := convertAWSMetadataOptionsToMAPI(fldPath.Child("instanceMetadataOptions"), m.awsMachine.Spec.InstanceMetadataOptions)
+	if err != nil {
+		errors = append(errors, errs...)
+	}
+
+	warnings = append(warnings, warn...)
+
+	mapaProviderConfig := mapiv1.AWSMachineProviderConfig{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "AWSMachineProviderConfig",
+			// In the machineSets both "awsproviderconfig.openshift.io/v1beta1" and "machine.openshift.io/v1beta1" can be found.
+			// Here we always settle on one of the two.
```

Comment:
This is just a string value, and yes, it could be an error in a round trip, because the object will have changed, but... I don't care 😅 

Nothing inspects the value of this field so it doesn't change anything functionally

Reply:
Ah - I meant for `APIVersrion`? Is that the same too?

Reply:
Yeah neither of these values actually matter, they just have to be populated so that RawExtension doesn't complain